core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala

```diff
@@ -0,0 +1,176 @@
+            "localhost", TaskLocality.NODE_LOCAL, false)
+          taskIndex += 1
+          // Track the successful commits in an accumulator. However, we can't just invoke
+          // accum += 1 since this unit test circumvents the usual accumulator updating
+          // infrastructure. So just send the accumulator update manually.
+          val accumUpdates = new mutable.HashMap[Long, Any]
+          try {
+            accumUpdates(accumId) = t.run(attemptNumber, attemptNumber)
+            dagSchedulerEventProcessLoop.post(
+              new CompletionEvent(t, Success, 0, accumUpdates, taskInfo, new TaskMetrics))
+          } catch {
+            case e: Throwable =>
+              dagSchedulerEventProcessLoop.post(new CompletionEvent(t, new ExceptionFailure(e,
+                Option.empty[TaskMetrics]), 1, accumUpdates, taskInfo, new TaskMetrics))
+          }
+        })
+      }
+    }
+
+    dagScheduler = new DAGScheduler(sc, taskScheduler)
+    taskScheduler.setDAGScheduler(dagScheduler)
+    sc.dagScheduler = dagScheduler
+    dagSchedulerEventProcessLoop = new DAGSchedulerSingleThreadedProcessLoop(dagScheduler)
+  }
+
+  /**
+   * Function that constructs a SparkHadoopWriter with a mock committer and runs its commit
+   */
+  private class OutputCommittingFunction
+      extends ((TaskContext, Iterator[Int]) => Int) with Serializable {
+
+    def apply(ctxt: TaskContext, it: Iterator[Int]): Int = {
+      val outputCommitter = new FakeOutputCommitter {
+        override def commitTask(taskAttemptContext: TaskAttemptContext) {
+          super.commitTask(taskAttemptContext)
+        }
+      }
+      runCommitWithProvidedCommitter(ctxt, it, outputCommitter)
+    }
+
+    protected def runCommitWithProvidedCommitter(
+        ctxt: TaskContext,
+        it: Iterator[Int],
+        outputCommitter: OutputCommitter): Int = {
+      def jobConf = new JobConf {
+        override def getOutputCommitter(): OutputCommitter = outputCommitter
+      }
+      val sparkHadoopWriter = new SparkHadoopWriter(jobConf) {
+        override def newTaskAttemptContext(
+            conf: JobConf,
+            attemptId: TaskAttemptID): TaskAttemptContext = {
+          mock(classOf[TaskAttemptContext])
+        }
+      }
+      sparkHadoopWriter.setup(ctxt.stageId, ctxt.partitionId, ctxt.attemptNumber)
+      sparkHadoopWriter.commit
+      if (FakeOutputCommitter.ran) {
+        FakeOutputCommitter.ran = false
+        1
+      } else {
+        0
+      }
+    }
+
+    @throws(classOf[IOException])
+    private def writeObject(out: ObjectOutputStream) {}
+
+    @throws(classOf[IOException])
+    private def readObject(in: ObjectInputStream) {}
+  }
+
+  /**
+   * Function that will explicitly fail to commit on the first attempt
+   */
+  private class FailFirstTimeCommittingFunction
+      extends OutputCommittingFunction {
+    override def apply(ctxt: TaskContext, it: Iterator[Int]): Int = {
+      if (ctxt.attemptNumber == 0) {
```

Comment:
This is a nice use of the new `attemptNumber` method (which is exactly why I merged that patch).  This sort of test is exactly why I'd also like to backport that patch to maintenance branches so that we can backport tests like this, too.